### ISSUE TITLE
S905: FP with optional chaining

### ISSUE THREAD

**Issue by quentin-jaquier-sonarsource** [SonarSource/SonarJS#1930]
Reproducer:

```ts
x?.somefunction("abc")
``` 

For S905, we don't have our own implementation, we are relying on "no-unused-expressions". The version of typescript-eslint we are currently using (we updated versions in #1849) already include the fix for this FP, if we use the implementation from typescript-eslint.
However, we are in fact [using a chai friendly](https://github.com/SonarSource/SonarJS/blob/fdc6425a982d78ef9dd62feccdad884ea52a4497/eslint-bridge/src/linter.ts#L58) implementation, who is based on eslint (therefore not supporting correctly TS...).

All in all, we are interested in the improvement of both implementations, but we are not supporting correctly optional chaining in the current situation.

**Comment by vilchik-elena** [comment]
Reported in https://community.sonarsource.com/t/typescript-3-7-2-null-conditional-operator-on-a-function-call-is-giving-a-false-positive/17791/4

**Comment by vilchik-elena** [comment]
Relates to #1931

**Comment by quentin-jaquier-sonarsource** [comment]
Reported in https://community.sonarsource.com/t/optional-chaining-in-typescript-is-reported-as-an-expression/21265

**Comment by andrea-guarino-sonarsource** [comment]
Reported in https://community.sonarsource.com/t/javascript-optional-chaining/23617

**Comment by dannymcgee** [comment]
Any updates on this? 6 months with no movement on a broken error check that's redundant with the TypeScript compiler's built-in functionality seems a little extreme. I don't see why this check couldn't just be removed entirely for TypeScript projects, but maybe there's something I'm missing?

**Comment by Crow-EH** [comment]
@dannymcgee Not a maintainer, but if Im not mistaken :
* SonarJS uses https://github.com/ihordiachenko/eslint-plugin-chai-friendly to check this rule
* eslint-plugin-chai-friendly is using ESLint only (without typescript-eslint) : https://github.com/ihordiachenko/eslint-plugin-chai-friendly/blob/master/package.json#L35
* ESLint alone does not support optional chaining yet

IMO there's 3 options for SonarJS :
* Contribute to / wait for ESLint then eslint-plugin-chai-friendly then update the dependency
* Find a way to add typescript-eslint to eslint-plugin-chai-friendly
* Stop using eslint-plugin-chai-friendly

Since eslint-plugin-chai-friendly is not maintained much anymore, I think dropping it is the best solution, but that's a lot of work and not for me to decide.

**Comment by dannymcgee** [comment]
I guess my other question would be why an "unused expression" is marked as a "bug" instead of something less severe like a "code smell." Short-circuited expressions (e.g. `foo && foo.functionCall()`) are super common in JavaScript, and I've never seen that pattern cause any errors or unexpected behavior. Whether you would prefer something more explicit like `if (foo) foo.functionCall();` is a matter of style/convention, not a "bug."

**Comment by thw0rted** [comment]
I've never had a short-circuited expression flagged as "unused" -- I mean, it's clearly used.  The problem is that, to a compiler that does not understand the optional-chaining operator, your `x.y?.z()` looks like `x.y`, i.e. an expression without side effects, and in 90% of cases it was probably a typo.  Maybe you meant `x.y()` or `x.y=z` or even

````js
a = b.
  x.y;
```

and you messed up the whitespace so that the automatic line continuation was wrong.  Point is, any *actual* "unused expression" almost certainly is a real bug.

**Comment by dannymcgee** [comment]
> I've never had a short-circuited expression flagged as "unused" -- I mean, it's clearly used.

Sorry for not being more clear; I looked up the eslint rule that's being used for this check and noticed that it has `allowShortCircuit` and `allowTernary` set to false by default. I assumed (probably incorrectly) that _that_ was the reason for the false positive here, and not just that the parser gets confused and reads the `?` as the end of the line.

**Comment by SantoJambit** [comment]
At what version is this fixed? we're at 6.2.2, which was released after this was closed, but it's not fixed in that version. And all later tags don't mention the fix.

**Comment by Crow-EH** [comment]
It's fixed in 6.3 ([according to this diff](https://github.com/SonarSource/SonarJS/compare/6.2.2.13315...6.3.0.12464))

**Comment by SantoJambit** [comment]
Thanks

**Comment by JuanCSoto** [comment]
Hello @Crow-EH.

I´m in Version 8.9.8 (build 54436) and i have the problem yet.

Thanks.

**Comment by Crow-EH** [comment]
Hi @JuanCSoto,

Again, not a maintainer, but please check this test and see if your case is tested : https://github.com/SonarSource/SonarJS/blob/master/eslint-bridge/tests/rules/no-unused-expressions.test.ts

If not, the test might be incomplete and should be completed.

Either way, since this issue is closed, for better visibility I recommend creating a new issue with a simple reproducer of your case, what you think should happen, and what actually happens.

Make sure it wasn't already reported of course.

Have a great evening !
